### PR TITLE
Di 22037 implement catalogs offline mode

### PR DESCRIPTION
**Pull request by saargt**
Move logic to a shared package

### CODE REVIEW

File: cpi-side/addon-cpi.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		throw new Error();
+	}
+}
+
+router.post('/:resourceName', async (req, res, next) => 
+{
+    
+	try 
+	{
+		// const genericResourceService = getGenericResourceService(req);
```
Comment (ShirMarciano) [inline cpi-side/addon-cpi.ts:41]: no implementation for post because we have catalog only for now?

File: cpi-side/addon-cpi.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	}
+});
+
+router.get('/:resourceName/key/:key', async (req, res, next) => 
+{
+    req.query.key = req.params.key;
+
+	try 
+	{
+		// const genericResourceService = getGenericResourceService(req);
```
Comment (ShirMarciano) [inline cpi-side/addon-cpi.ts:59]: No need the validation here? as you did above?

File: shared/src/baseCore.service.ts [hunk 3]
```diff
@@ -369,7 +357,7 @@ export class BaseCoreService
 		res.DIMXObjects = papiBatchResult.map(papiItem => 
 		{
 			return {
-				Key: papiItem.UUID,
+				Key: papiItem.UUID!,
```
Comment (ShirMarciano) [inline shared/src/baseCore.service.ts:360]: why the '!'?

File: cpi-side/addon-cpi.ts (hunk 1 above)
Comment (saargt) [reply cpi-side/addon-cpi.ts:41]: CPI-side is currently just a copy-paste. Nothing's really implemented.

File: cpi-side/addon-cpi.ts (hunk 2 above)
Comment (saargt) [reply cpi-side/addon-cpi.ts:59]: Same here.

File: shared/src/baseCore.service.ts (hunk 3 above)
Comment (saargt) [reply shared/src/baseCore.service.ts:360]: That says "papiItem will have a UUID, trust me on this Typescript.'

**Review by ShirMarciano — APPROVED** [review]

File: cpi-side/clientApiService.ts [hunk 4]
```diff
@@ -55,6 +56,100 @@ export default class ClientApiService implements IPapiService
...
+            "Objects": clientApiSearchResult.objects,
+            ...(body.IncludeCount && {"Count" : clientApiSearchResult.count})
+        };
+
+		return searchResult!;
+    }
+
+    async handleSearchUniqueFieldList(resourceName: string, uniqueField: string, valuesList: string[], fields: string[], page: number = 0, pageSize?: number): Promise<ClientApiSearchResult<string>>
+    {
+        debugger;
```
Comment (ShirMarciano) [inline cpi-side/clientApiService.ts]: please remove:)

File: cpi-side/clientApiService.ts [hunk 5]
```diff
@@ -170,24 +176,15 @@ export default class ClientApiService implements IPapiService
         const requestedPageResourceWithFields: Array<any> = [];
         for (const resource of requestedPageResources)
         {
-            requestedPageResourceWithFields.push(_.pick(resource, fields));
+            // For explanation about lodash.pick, see: https://lodash.com/docs/4.17.15#pick
```
Comment (ShirMarciano) [inline cpi-side/clientApiService.ts]: Thanks for that

**Pull request merged**